docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst

```diff
@@ -0,0 +1,318 @@
+
+.. code-block:: ql
+
+    import python
+    import semmle.python.dataflow.new.DataFlow
+    import semmle.python.ApiGraphs
+
+    from DataFlow::CallCfgNode call, DataFlow::ExprNode expr
+    where
+      call = API::moduleImport("os").getMember("open").getACall() and
+      DataFlow::localFlow(expr, call.getArg(0))
+    select call, expr
+
+➤ `See this in the query console on LGTM.com <https://lgtm.com/query/8213643003890447109/>`__. Many expressions flow to the same call.
+
+We see that we get several data-flow nodes for an expression as it flows towards a call (notice repeated locations in the ``call`` column). We are mostly interested in the "first" of these, what might be called the local source for the file name. To restrict attention to such local sources, and to simultaneously make the analysis more performant, we have the QL class ``LocalSourceNode``:
+
+.. code-block:: ql
+
+    import python
+    import semmle.python.dataflow.new.DataFlow
+    import semmle.python.ApiGraphs
+
+    from DataFlow::CallCfgNode call, DataFlow::ExprNode expr
+    where
+      call = API::moduleImport("os").getMember("open").getACall() and
+      DataFlow::localFlow(expr, call.getArg(0)) and
+      expr instanceof DataFlow::LocalSourceNode
+    select call, expr
+
+➤ `See this in the query console on LGTM.com <https://lgtm.com/query/2017139821928498055/>`__. We now mostly have one expression per call.
+
+We still have some cases of more than one expression flowing to a call, but then they flow through different code paths (possibly due to control-flow splitting, as in the second case).
+
+We can also make the source more specific, for example a parameter to a function or method. This query finds instances where a parameter is used as the name when opening a file:
+
+.. code-block:: ql
+
+    import python
+    import semmle.python.dataflow.new.DataFlow
+    import semmle.python.ApiGraphs
+
+    from DataFlow::CallCfgNode call, DataFlow::ParameterNode p
+    where
+      call = API::moduleImport("os").getMember("open").getACall() and
+      DataFlow::localFlow(p, call.getArg(0))
+    select call, p
+
+➤ `See this in the query console on LGTM.com <https://lgtm.com/query/3998032643497238063/>`__. Very few hits now; these could feasibly be inspected manually.
+
+Using the exact name supplied via the parameter may be too strict. If we want to know if the parameter influences the file name, we can use taint tracking instead of data flow. This query finds calls to ``os.open`` where the filename is derived from a parameter:
+
+.. code-block:: ql
+
+    import python
+    import semmle.python.dataflow.new.TaintTracking
+    import semmle.python.ApiGraphs
+
+    from DataFlow::CallCfgNode call, DataFlow::ParameterNode p
+    where
+      call = API::moduleImport("os").getMember("open").getACall() and
+      TaintTracking::localTaint(p, call.getArg(0))
```

Comment:
In light of my comment in the data-flow version of this example, I wonder if we should have `taintFlowsTo` as a method on `LocalSourceNode` and perhaps discourage the use of `localFlow`/`localTaint` entirely.

Reply:
That might be good.